openembedded-core-mbl/scripts/lib/mbl-signing/signing/pki/x509utils.py

```diff
@@ -0,0 +1,315 @@
+        )
+        self.value = enc_val
+
+
+class PKExtension(ATFExtension):
+    """ATFExtension containing a SubjectPublicKeyInfo value."""
+
+    def __init__(self, *args, pub_key_name=None, **kwargs):
+        """
+        Init the extension object.
+
+        :params pub_key_name str: The name of the public key to
+        include in the extension.
+        """
+        super().__init__(*args, **kwargs)
+        self.pub_key_name = pub_key_name
+
+    def encode_value(self, pub_key):
+        """
+        Encode a subject public key and store in the extension.
+
+        :params pub_key bytes: The raw subject public key bytes.
+        """
+        enc_val = asn1parse.encode_asn1_subject_key_info(
+            "1.2.840.113549.1.1.1", pub_key
+        )
+        self.value = enc_val
+
+
+class CounterExtension(ATFExtension):
+    """ATFExtension containing NV counter values."""
+
+    def encode_value(self, counter_value):
+        """Encode the counter value as an asn1 integer."""
+        self.value = asn1parse.encode_asn1_nv_counter_info(counter_value)
+
+
+class X509CertificateExtensions:
+    """A collection of ATFExtensions to be included in a certificate."""
+
+    def __init__(self, name, exts, signer, file_name):
+        """
+        Initialise the object.
+
+        :params name str: Certificate common name.
+        :params exts dict: Dict of extension types to include.
+        :params signer str: The signing key name.
+        :params file_name str: Output file name for the cert.
+        """
+        self._extensions = exts
+        self._cert_name = name
+        self._signer = signer
+        self._file_name = file_name
+
+    @property
+    def name(self):
+        """Return the certificate name."""
+        return self._cert_name
+
+    @property
+    def file_name(self):
+        """Return the output file name."""
+        return self._file_name
+
+    @property
+    def signer(self):
+        """Return the signer name."""
+        return self._signer
+
+    @property
+    def extensions(self):
+        """Return a list of extension objects."""
+        return list(self._extensions.values())
+
+    def get_extension(self, name):
+        """Get an extension by name."""
+        return self._extensions[name]
+
+    def get_ext_name_from_oid(self, oid):
+        """Return the extension name associated with an OID."""
+        for name, ext in self._extensions.items():
+            if ext.oid == oid:
+                return name
+
+    def add_extension(self, name, ext):
+        """Add a new extension to the set."""
+        self._extensions[name] = ext
+
+    def __iter__(self):
+        """Make the object iterable so we can iterate over the extensions."""
+        for i in self._extensions.values():
+            yield i
+
+
+class ExtensionValueGetter:
+    """
+    This thing deduces the extension value type.
+
+    It does this based on the extension object's type.
+    This is an abominable hack and should be removed.
```

Comment:
Do we have an idea of what we should be doing instead? Do we have a ticket/task for this?
It would be better to explain this is a work-around until we have implemented X.

Reply:
I'm planning on improving this code when I come to work on the release version of the signing tool. I'll add it as a point to one of the tasks. It's not really a work around, it's just a bit rubbish, but I don't have time to come up with anything better than this right now.

Reply:
Changed the docstring in 712d58f